torchmetrics/detection/map_new.py

```diff
@@ -0,0 +1,600 @@
+            if iouThr is not None:
+                t = torch.where(iouThr == self.iou_thresholds)[0]
+                s = s[t]
+            s = s[:, :, :, aind, mind]
+        else:
+            # dimension of recall: [TxKxAxM]
+            s = results["recall"]
+            if iouThr is not None:
+                t = torch.where(iouThr == self.iou_thresholds)[0]
+                s = s[t]
+            s = s[:, :, aind, mind]
+        if len(s[s > -1]) == 0:
+            mean_s = torch.Tensor([-1])
+        else:
+            mean_s = torch.mean(s[s > -1])
+
+        return mean_s
+
+    def _calculate(self, classes) -> Tuple[Dict, MAPMetricResults, MARMetricResults]:
+        img_ids = torch.arange(len(self.groundtruth_boxes), dtype=torch.int).tolist()
+        imgIds = img_ids
+        catIds = classes
+
+        maxDets = self.max_detection_thresholds
+        maxDet = self.max_detection_thresholds[-1]
+        areaRng = self.object_area_ranges.values()
+
+        ious = {(imgId, catId): self._compute_iou(imgId, catId, maxDet) for imgId in imgIds for catId in catIds}
+
+        evalImgs = [
+            self._evaluate_image(imgId, catId, area, maxDet, ious)
+            for catId in catIds
+            for area in areaRng
+            for imgId in imgIds
+        ]
+
+        T = len(self.iou_thresholds)
+        R = len(self.rec_thresholds)
+        K = len(catIds)
+        A = len(self.object_area_ranges)
+        M = len(self.max_detection_thresholds)
+        precision = -torch.ones((T, R, K, A, M))
+        recall = -torch.ones((T, K, A, M))
+        scores = -torch.ones((T, R, K, A, M))
+
+        setK = set(catIds)
+        setA = set(map(tuple, self.object_area_ranges.values()))
+        setM = set(self.max_detection_thresholds.tolist())
+        setI = set(imgIds)
+        # get inds to evaluate
+        k_list = [n for n, k in enumerate(catIds) if k in setK]
+        m_list = [m for n, m in enumerate(maxDets.tolist()) if m in setM]
+        a_list = [n for n, a in enumerate(map(lambda x: tuple(x), areaRng)) if a in setA]
+        i_list = [n for n, i in enumerate(imgIds) if i in setI]
+        I0 = len(imgIds)
+        A0 = len(areaRng)
+        # # retrieve E at each category, area range, and max number of detections
+        for k, k0 in enumerate(k_list):
+            Nk = k0 * A0 * I0
+            for a, a0 in enumerate(a_list):
+                Na = a0 * I0
+                for m, maxDet in enumerate(m_list):
+                    E = [evalImgs[Nk + Na + i] for i in i_list]
+                    E = [e for e in E if e is not None]
+                    if len(E) == 0:
+                        continue
+                    dtScores = torch.cat([e["dtScores"][:maxDet] for e in E])
+
+                    # different sorting method generates slightly different results.
+                    # mergesort is used to be consistent as Matlab implementation.
+                    inds = torch.argsort(dtScores, descending=True)
```

Comment:
This seems to be a problem.

Reply:
Weird that different sorting has an impact... Do you know how big it is?

Reply:
I'm not sure, it actually has a big impact, but I cant validate yet, cause there is still an issue inside the `_evaluate_image` function that I have to fix tomorrow

Reply:
Tests pass, so let's hope, this works for now